### DISCUSSION TITLE
Building a component library with the ability to override inner styles from outside

### DISCUSSION THREAD

**Discussion by sedghi** [tailwindlabs/tailwindcss#9343] [Help]
I'm a core developer for one of the most popular open source medical imaging application on the web called [OHIF Viewer](https://github.com/OHIF/Viewers/tree/v3-stable), you can take a look at our demo [here](https://v3-demo.ohif.org/) built using React and TailwindCSS. You can look at our component library [here](https://v3-ui.ohif.org/) as well.

During the upgrade to TailwindCSS 3.x (to get all the new good stuff), I saw a lot of broken components. Digging deeper, I found out that TailwindCSS 3.x does not override the previous styles if defined before which through out the app we are relying on them. 

Let me give you an example, so for instance, our Button component is like below (simplified)

```js
const baseClasses =
  'leading-none font-sans text-center justify-center items-center outline-none transition duration-300 ease-in-out focus:outline-none';

const variantClasses = {
  outlined: {
      default:
        'border bg-trasparent border-primary-light text-primary-light hover:bg-primary-light hover:text-black focus:text-black focus:bg-primary-light active:opacity-80',
      primary:
        'border bg-transparent border-primary-main text-primary-main hover:opacity-80 active:opacity-100 focus:opacity-80',
      secondary:
        'border bg-transparent border-secondary-light text-secondary-light hover:opacity-80 active:opacity-100 focus:opacity-80',
      white:
        'border bg-transparent border-white text-white hover:opacity-80 active:opacity-100 focus:opacity-80',
      black:
        'border bg-black border-primary-main text-white hover:bg-primary-main focus:bg-primary-main hover:border-black focus:border-black',
    },
};

/** some more variants here */

const Button = ({
  children,
  variant = defaults.variant,
  color = defaults.color,
  size = defaults.size,
  rounded = defaults.rounded,
  disabled = defaults.disabled,
  type = defaults.type,
  fullWidth = defaults.fullWidth,
  startIcon: startIconProp,
  endIcon: endIconProp,
  name,
  className,
  onClick,
  ...rest
}) => {

  return (
    <button
      className={classnames(
        className,
        baseClasses,
        variantClasses[variant][color],
        roundedClasses[rounded],
        sizeClasses[size],
      )}
    >
      {children}
    </button>
  );
};
```

and then we use it like 

```js
<Button variant="outlined" onClick={onResetHandler}>
     Reset to Defaults'
</Button>
```

However, one can add `classNames` to the Button as well (which will get appended to the button styles), like 

```js
<Button
  variant="outlined"
  className="border-primary-main px-4 py-2 text-base"
  color="white"
>
  {`<<`}
</Button>

```

However, as I said, in Tailwind 3.x the `border-primary-main` on the className would not necessary get applied since there is already some other border stylings happening using the `contained` and `white` variants (`border-white`).
So basically, if one wants all the stylings of the `outline` and `white` variant but wants to override only the `border-white` they can't with our current design of our components.

I have looked around this problem and basically there are two options. 1) Using some library like [tailwind merge](https://www.npmjs.com/package/tailwind-merge) to merge the styles which I don't like because of maintainability concerns OR 2) explicitly defining each style via props which would lead to a lot of props ...

I was wondering how would one approach this problem of overriding the component styles. Thanks in advance 

Some other links I have found: 
- https://twitter.com/adamwathan/status/1496663833980686338

**Comment by wongjn** [comment]
The options you already looked at are the only options really – use a class management library or use more props.

**Comment by AssisrMatheus** [comment]
What if you change

```js
className={classnames(
        className,
        baseClasses,
        variantClasses[variant][color],
        roundedClasses[rounded],
        sizeClasses[size],
      )}
```

to:

```js
className={classnames(
        baseClasses,
        variantClasses[variant][color],
        roundedClasses[rounded],
        sizeClasses[size],
        className
      )}
```

**Reply by n18l** [reply]
This is a reasonable solution for other props in React that you might need to override, but Tailwind classes are a little different.

What this will do is append the user-provided classes last—but it won't _replace_ any preexisting classes that affect the same properties. When these types of conflicting classes exist on a single element, the "Cascading" nature of CSS takes over and the properties of whichever class comes later in the stylesheet win out. The order of the classes in the HTML doesn't actually matter at all.

When Tailwind generates your stylesheet, it doesn't guarantee the rules are in any particular order. In fact, its nature as a utility library means it really doesn't want to. It instead relies on the developer to make sure there aren't any conflicting classes on a given element, which is one of the problems packages like [tailwind-merge](https://www.npmjs.com/package/tailwind-merge) try to solve.

**Comment by jamesarosen** [comment]
You could ensure that the default styles appear before the overrides in the emitted CSS by putting the defaults into the `components` layer:
```css
@layer components {
  .OHIF-Button {
    @apply leading-none font-sans text-center justify-center items-center outline-none transition duration-300 ease-in-out focus:outline-none;
  }
}
```
That way, `.border-primary-main` would win in the CSS.
```js
<Button className="border-primary-main">
```
would still have both classes, though, so there would be some CSS bloat.

**Comment by paulwongx** [comment]
I think many of us have this problem and many folks use clsx and tailwind-merge in a combo function to override styles. Works great frankly.
